File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/publicALaCarteProducts/PublicAlaCarteProductsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.publicALaCarteProducts

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.publicALaCarteProducts.PublicAlaCarteProductsApi
import features.publicALaCarteProducts.models.PublicALaCarteProductsModel

import features.publicALaCarteProducts.serializePublicAlaCarteProductsState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class PublicAlaCarteProductsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_public_ala_carte_products"

    override val eventChannel: String
        get() = "customer_sdk/events/get_public_ala_carte_products"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getPublicAlaCarteProducts(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getPublicAlaCarteProducts(call: MethodCall): String? {
            val result = PublicAlaCarteProductsApi().getPublicAlaCarteProducts().toJsonResult<PublicALaCarteProductsModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        PublicAlaCarteProductsApi().publicAlaCarteProductsState.serializePublicAlaCarteProductsState().observe { result ->
            pushSerializedResultToEventSink<PublicAlaCarteProductsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/publicALaCarteProducts/PublicAlaCarteProductsHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class PublicAlaCarteProductsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_public_ala_carte_products"

    var eventChannel: String = "customer_sdk/events/get_public_ala_carte_products"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await PublicAlaCarteProductsHandler.getPublicAlaCarteProducts(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getPublicAlaCarteProducts(call: FlutterMethodCall) async -> String? {
        do {
        let result = try await PublicAlaCarteProductsApi().getPublicAlaCarteProducts().serializePublicAlaCarteProductsApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
            }
        } catch {
            return nil
        }

        


        

        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        PublicAlaCarteProductsApi().publicAlaCarteProductsState
            .serializePublicAlaCarteProductsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
